basicsr/archs/hifacegan_arch.py

```diff
@@ -0,0 +1,284 @@
+
+
+@ARCH_REGISTRY.register()
+class SPADEGenerator(BaseNetwork):
+    """Generator with SPADEResBlock"""
+
+    def __init__(self,
+                 num_in_ch=3,
+                 num_feat=64,
+                 use_vae=False,
+                 z_dim=256,
+                 num_upsampling_layers='normal',
+                 crop_size=512,
+                 norm_G='spectralspadesyncbatch3x3',
+                 is_train=True,
+                 init_train_phase=3):  # progressive training disabled
+        super().__init__()
+        self.nf = num_feat
+        self.input_nc = num_in_ch
+        self.is_train = is_train
+        self.train_phase = init_train_phase
+
+        if num_upsampling_layers == 'normal':
+            num_up_layers = 5
+        elif num_upsampling_layers == 'more':
+            num_up_layers = 6
+        elif num_upsampling_layers == 'most':
+            num_up_layers = 7
+        else:
+            raise ValueError(f'opt.num_upsampling_layers [{num_upsampling_layers}] not recognized')
+
+        # 20200211 Yang Lingbo with respect to phase
+        self.scale_ratio = num_up_layers
+        self.sw = crop_size // (2**num_up_layers)
+        self.sh = self.sw  # 20210519: By default use square image, aspect_ratio = 1.0
+
+        if use_vae:
+            # In case of VAE, we will sample from random z vector
+            self.fc = nn.Linear(z_dim, 16 * self.nf * self.sw * self.sh)
+        else:
+            # Otherwise, we make the network deterministic by starting with
+            # downsampled segmentation map instead of random z
+            self.fc = nn.Conv2d(num_in_ch, 16 * self.nf, 3, padding=1)
+
+        self.head_0 = SPADEResnetBlock(16 * self.nf, 16 * self.nf, norm_G)
+
+        self.G_middle_0 = SPADEResnetBlock(16 * self.nf, 16 * self.nf, norm_G)
+        self.G_middle_1 = SPADEResnetBlock(16 * self.nf, 16 * self.nf, norm_G)
+
+        self.ups = nn.ModuleList([
+            SPADEResnetBlock(16 * self.nf, 8 * self.nf, norm_G),
+            SPADEResnetBlock(8 * self.nf, 4 * self.nf, norm_G),
+            SPADEResnetBlock(4 * self.nf, 2 * self.nf, norm_G),
+            SPADEResnetBlock(2 * self.nf, 1 * self.nf, norm_G)  # here
+        ])
+
+        self.to_rgbs = nn.ModuleList([
+            nn.Conv2d(8 * self.nf, 3, 3, padding=1),
+            nn.Conv2d(4 * self.nf, 3, 3, padding=1),
+            nn.Conv2d(2 * self.nf, 3, 3, padding=1),
+            nn.Conv2d(1 * self.nf, 3, 3, padding=1)  # here
+        ])
+
+        self.up = nn.Upsample(scale_factor=2)
+
+    # 20200309 interface for flexible encoder design
+    # and mid-level loss control!
+    # For basic network, it's just a 16x downsampling
+    def encode(self, input_tensor):
+        h, w = input_tensor.size()[-2:]
+        sh, sw = h // 2**self.scale_ratio, w // 2**self.scale_ratio
+        x = F.interpolate(input_tensor, size=(sh, sw))
+        return self.fc(x)  # 20200310: Merge fc into encoder
```

Comment:
If it is determined, I think it is better to remove the date~ 
And I think this comment is a bit redundant？

Reply:
yeah, all legacy code~